[jit] use fully qualified name for ScriptClasses

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #19575 [jit] Use QualifiedName for classes
* #19626 [jit] make QualifiedName a value type
* #19566 [jit] separate QualifiedName into its own file
* **#19239 [jit] use fully qualified name for ScriptClasses**

Make the necessary changes so that we use fully qualified names for
classes:
- Classes are registered by qualified name.
- PythonPrintPass prints qualified names for classes, and prints import
statements as well so that we can resolve those class names.
- Give ScriptTypeParser the ability to resolve qualified names to types.
- Give compiler.cpp the ability to resolve qualified names to types with
a class namespace SugaredValue.
- Change the lib export file structure to match Python, e.g. the class
foo.bar.Baz ends up in `class Baz` in foo/bar.py.

gh-metadata: pytorch pytorch 19239 gh/suo/27/head

Differential Revision: [D14928293](https://our.internmc.facebook.com/intern/diff/D14928293)

suo committed:
[jit] use fully qualified name for ScriptClasses

suo committed:
Update on "[jit] use fully qualified name for ScriptClasses"

[jit] use fully qualified name for ScriptClasses

gh-metadata: pytorch pytorch 19239 gh/suo/27/head

suo committed:
Update on "[jit] use fully qualified name for ScriptClasses"

[jit] use fully qualified name for ScriptClasses

Make the necessary changes so that we use fully qualified names for
classes:
- Classes are registered by qualified name.
- PythonPrintPass prints qualified names for classes, and prints import
statements as well so that we can resolve those class names.
- Give ScriptTypeParser the ability to resolve qualified names to types.
- Give compiler.cpp the ability to resolve qualified names to types with
a class namespace SugaredValue.
- Change the lib export file structure to match Python, e.g. the class
foo.bar.Baz ends up in `class Baz` in foo/bar.py.

gh-metadata: pytorch pytorch 19239 gh/suo/27/head

suo committed:
Update base for Update on "[jit] use fully qualified name for ScriptClasses"

[jit] use fully qualified name for ScriptClasses

Make the necessary changes so that we use fully qualified names for
classes:
- Classes are registered by qualified name.
- PythonPrintPass prints qualified names for classes, and prints import
statements as well so that we can resolve those class names.
- Give ScriptTypeParser the ability to resolve qualified names to types.
- Give compiler.cpp the ability to resolve qualified names to types with
a class namespace SugaredValue.
- Change the lib export file structure to match Python, e.g. the class
foo.bar.Baz ends up in `class Baz` in foo/bar.py.

gh-metadata: pytorch pytorch 19239 gh/suo/27/head

suo committed:
Update base for Update on "[wip][jit] use fully qualified name for ScriptClasses"

[jit] use fully qualified name for ScriptClasses

Make the necessary changes so that we use fully qualified names for
classes:
- Classes are registered by qualified name.
- PythonPrintPass prints qualified names for classes, and prints import
statements as well so that we can resolve those class names.
- Give ScriptTypeParser the ability to resolve qualified names to types.
- Give compiler.cpp the ability to resolve qualified names to types with
a class namespace SugaredValue.
- Change the lib export file structure to match Python, e.g. the class
foo.bar.Baz ends up in `class Baz` in foo/bar.py.

gh-metadata: pytorch pytorch 19239 gh/suo/27/head

suo committed:
Update on "[wip][jit] use fully qualified name for ScriptClasses"

[jit] use fully qualified name for ScriptClasses

Make the necessary changes so that we use fully qualified names for
classes:
- Classes are registered by qualified name.
- PythonPrintPass prints qualified names for classes, and prints import
statements as well so that we can resolve those class names.
- Give ScriptTypeParser the ability to resolve qualified names to types.
- Give compiler.cpp the ability to resolve qualified names to types with
a class namespace SugaredValue.
- Change the lib export file structure to match Python, e.g. the class
foo.bar.Baz ends up in `class Baz` in foo/bar.py.

gh-metadata: pytorch pytorch 19239 gh/suo/27/head

suo committed:
Update on "[wip][jit] use fully qualified name for ScriptClasses"

[jit] use fully qualified name for ScriptClasses

Make the necessary changes so that we use fully qualified names for
classes:
- Classes are registered by qualified name.
- PythonPrintPass prints qualified names for classes, and prints import
statements as well so that we can resolve those class names.
- Give ScriptTypeParser the ability to resolve qualified names to types.
- Give compiler.cpp the ability to resolve qualified names to types with
a class namespace SugaredValue.
- Change the lib export file structure to match Python, e.g. the class
foo.bar.Baz ends up in `class Baz` in foo/bar.py.

gh-metadata: pytorch pytorch 19239 gh/suo/27/head

suo committed:
Update on "[wip][jit] use fully qualified name for ScriptClasses"

[jit] use fully qualified name for ScriptClasses

Make the necessary changes so that we use fully qualified names for
classes:
- Classes are registered by qualified name.
- PythonPrintPass prints qualified names for classes, and prints import
statements as well so that we can resolve those class names.
- Give ScriptTypeParser the ability to resolve qualified names to types.
- Give compiler.cpp the ability to resolve qualified names to types with
a class namespace SugaredValue.
- Change the lib export file structure to match Python, e.g. the class
foo.bar.Baz ends up in `class Baz` in foo/bar.py.

gh-metadata: pytorch pytorch 19239 gh/suo/27/head

suo committed:
Update on "[wip][jit] use fully qualified name for ScriptClasses"

[jit] use fully qualified name for ScriptClasses

Make the necessary changes so that we use fully qualified names for
classes:
- Classes are registered by qualified name.
- PythonPrintPass prints qualified names for classes, and prints import
statements as well so that we can resolve those class names.
- Give ScriptTypeParser the ability to resolve qualified names to types.
- Give compiler.cpp the ability to resolve qualified names to types with
a class namespace SugaredValue.
- Change the lib export file structure to match Python, e.g. the class
foo.bar.Baz ends up in `class Baz` in foo/bar.py.

gh-metadata: pytorch pytorch 19239 gh/suo/27/head

suo committed:
Update on "[wip][jit] use fully qualified name for ScriptClasses"

[jit] use fully qualified name for ScriptClasses

Make the necessary changes so that we use fully qualified names for
classes:
- Classes are registered by qualified name.
- PythonPrintPass prints qualified names for classes, and prints import
statements as well so that we can resolve those class names.
- Give ScriptTypeParser the ability to resolve qualified names to types.
- Give compiler.cpp the ability to resolve qualified names to types with
a class namespace SugaredValue.
- Change the lib export file structure to match Python, e.g. the class
foo.bar.Baz ends up in `class Baz` in foo/bar.py.

gh-metadata: pytorch pytorch 19239 gh/suo/27/head

suo committed:
Update on "[wip][jit] use fully qualified name for ScriptClasses"

[jit] use fully qualified name for ScriptClasses

Make the necessary changes so that we use fully qualified names for
classes:
- Classes are registered by qualified name.
- PythonPrintPass prints qualified names for classes, and prints import
statements as well so that we can resolve those class names.
- Give ScriptTypeParser the ability to resolve qualified names to types.
- Give compiler.cpp the ability to resolve qualified names to types with
a class namespace SugaredValue.
- Change the lib export file structure to match Python, e.g. the class
foo.bar.Baz ends up in `class Baz` in foo/bar.py.

gh-metadata: pytorch pytorch 19239 gh/suo/27/head

suo committed:
Update on "[wip][jit] use fully qualified name for ScriptClasses"

[jit] use fully qualified name for ScriptClasses

Make the necessary changes so that we use fully qualified names for
classes:
- Classes are registered by qualified name.
- PythonPrintPass prints qualified names for classes, and prints import
statements as well so that we can resolve those class names.
- Give ScriptTypeParser the ability to resolve qualified names to types.
- Give compiler.cpp the ability to resolve qualified names to types with
a class namespace SugaredValue.
- Change the lib export file structure to match Python, e.g. the class
foo.bar.Baz ends up in `class Baz` in foo/bar.py.

gh-metadata: pytorch pytorch 19239 gh/suo/27/head

suo committed:
fix py2 on "[wip][jit] use fully qualified name for ScriptClasses"

[jit] use fully qualified name for ScriptClasses

Make the necessary changes so that we use fully qualified names for
classes:
- Classes are registered by qualified name.
- PythonPrintPass prints qualified names for classes, and prints import
statements as well so that we can resolve those class names.
- Give ScriptTypeParser the ability to resolve qualified names to types.
- Give compiler.cpp the ability to resolve qualified names to types with
a class namespace SugaredValue.
- Change the lib export file structure to match Python, e.g. the class
foo.bar.Baz ends up in `class Baz` in foo/bar.py.

gh-metadata: pytorch pytorch 19239 gh/suo/27/head

suo committed:
fix lint on "[wip][jit] use fully qualified name for ScriptClasses"

[jit] use fully qualified name for ScriptClasses

Make the necessary changes so that we use fully qualified names for
classes:
- Classes are registered by qualified name.
- PythonPrintPass prints qualified names for classes, and prints import
statements as well so that we can resolve those class names.
- Give ScriptTypeParser the ability to resolve qualified names to types.
- Give compiler.cpp the ability to resolve qualified names to types with
a class namespace SugaredValue.
- Change the lib export file structure to match Python, e.g. the class
foo.bar.Baz ends up in `class Baz` in foo/bar.py.

gh-metadata: pytorch pytorch 19239 gh/suo/27/head

suo committed:
Update on "[wip][jit] use fully qualified name for ScriptClasses"

[jit] use fully qualified name for ScriptClasses

Make the necessary changes so that we use fully qualified names for
classes:
- Classes are registered by qualified name.
- PythonPrintPass prints qualified names for classes, and prints import
statements as well so that we can resolve those class names.
- Give ScriptTypeParser the ability to resolve qualified names to types.
- Give compiler.cpp the ability to resolve qualified names to types with
a class namespace SugaredValue.
- Change the lib export file structure to match Python, e.g. the class
foo.bar.Baz ends up in `class Baz` in foo/bar.py.

gh-metadata: pytorch pytorch 19239 gh/suo/27/head

suo committed:
Update on "[wip][jit] use fully qualified name for ScriptClasses"

[jit] use fully qualified name for ScriptClasses

Make the necessary changes so that we use fully qualified names for
classes:
- Classes are registered by qualified name.
- PythonPrintPass prints qualified names for classes, and prints import
statements as well so that we can resolve those class names.
- Give ScriptTypeParser the ability to resolve qualified names to types.
- Give compiler.cpp the ability to resolve qualified names to types with
a class namespace SugaredValue.
- Change the lib export file structure to match Python, e.g. the class
foo.bar.Baz ends up in `class Baz` in foo/bar.py.

gh-metadata: pytorch pytorch 19239 gh/suo/27/head

suo committed:
Update on "[wip][jit] use fully qualified name for ScriptClasses"

[jit] use fully qualified name for ScriptClasses

Make the necessary changes so that we use fully qualified names for
classes:
- Classes are registered by qualified name.
- PythonPrintPass prints qualified names for classes, and prints import
statements as well so that we can resolve those class names.
- Give ScriptTypeParser the ability to resolve qualified names to types.
- Give compiler.cpp the ability to resolve qualified names to types with
a class namespace SugaredValue.
- Change the lib export file structure to match Python, e.g. the class
foo.bar.Baz ends up in `class Baz` in foo/bar.py.

gh-metadata: pytorch pytorch 19239 gh/suo/27/head

zdevito commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1319,12 +1319,25 @@ struct CAFFE2_API ClassType : public Type {
+    return qualifiedName_;
+  }
+
+  const std::string& qualname() const {
```
We have a QualifiedName object in python_print.cpp that should be used here. Otherwise, expect endless bugs about passing an unqualified name when a qualified one was expected and vice-versa. It also will keep name parsing code in the parser.

zdevito commented on caffe2/proto/torch.proto:
```diff
@@ -104,5 +104,5 @@ message ModelDef {
   // TorchScript code that this module hierarchy depends on
   // Libs will be loaded in before the main module is compiled.
-  optional LibDef libs = 9;
+  repeated LibDef libs = 9;
```
I think this should just list the directory root where these files live. I am inclined to avoid duplication of this information because it makes it easier to hack on the file for testing purposes.

zdevito replied:
Because this is a change to the proto version, I am apprehensive about spreading it across multiple commits.

suo replied:
Makes sense. I guess we don't even really need a field in this case, since we can just consider `libs/` the root.

zdevito commented on torch/csrc/jit/export.cpp:
```diff
@@ -580,23 +582,47 @@ void ScriptModuleSerializer::serialize(
+        ImportExportHelpers::qualifierToPath(class_type->qualifier());
+    // End state: filename is "foo/bar.py", in which we will define a class
+    // named Baz
+    fileToSrc[filename] += class_src;
```
fileToSrc should be string -> ostringstream, otherwise this is N^2.

zdevito commented on torch/csrc/jit/export.cpp:
```diff
@@ -580,23 +582,47 @@ void ScriptModuleSerializer::serialize(
-  const auto& lib_str = lib_stream.str();
-  writer_.writeRecord(filename, lib_str.c_str(), lib_str.size());
-  lib_record->set_key(filename);
+  // Write out the files. We still have to do this in converted_classes_ order,
```
Why? Isn't that what the import statements do?

zdevito commented on torch/csrc/jit/import_export_helpers.cpp:
```diff
@@ -0,0 +1,42 @@
+static const std::string kExportPrefix = "libs/";
+static const std::string kExportSuffix = "py";
+
+std::string qualifierToPath(const std::string& qualifier) {
```
These really should be using the QualifiedName type.

zdevito commented on torch/csrc/jit/import_export_helpers.cpp:
```diff
@@ -0,0 +1,42 @@
+
+namespace torch {
+namespace jit {
+namespace ImportExportHelpers {
```
Do we really need a new namespace? Its style doesn't match namespace style.

suo replied:
I am basically trying to accomplish the same thing as a class with only static methods.

zdevito commented on torch/csrc/jit/export.cpp:
```diff
@@ -580,23 +582,47 @@ void ScriptModuleSerializer::serialize(
+    const std::string& src = fileToSrc.at(filename);
+
+    std::ostringstream lib_stream;
+    lib_stream << "op_version_set = " << op_version_set << "\n";
```
aside: we should move the op_version_set into the json file, now that there are so many source files. This would also clear up a bunch of tests where we have to manually prepend this thing.

zdevito commented on torch/csrc/jit/script/sugared_value.h:
```diff
@@ -350,6 +343,30 @@ static inline Self simpleSelf(const TypePtr& typ) {
   };
 }
 
+// Represents nested class namespaces, like `foo.bar.Baz`.
```
Can you put this with the other import-specific sugared values in import_source.cpp? Otherwise people will get confused and try to use there in other places.

zdevito commented on torch/jit/__init__.py:
```diff
@@ -727,6 +727,26 @@ def _is_new_style_class(cls):
 
 
+# Retrieves a fully-qualified name (module hierarchy + classname) for a given obj.
+def _qualified_name(obj):
```
I believe there should also be a check: 
```getattr(lookup_module(module_name), name) is obj```

Otherwise bad things can happen if someone tries to just set __module__. This is how the pickler works as well.

suo replied:
This unfortunately doesn't work with nested classes (or classes defined in functions), since python pre-3.3 doesn't have a way of figuring out the nested name within a module.

So adding this check makes writing the tests very tricky, since we have to un-nest everything. I'd prefer to make that change in a follow-up since this PR is big already.

zdevito commented on torch/jit/__init__.py:
```diff
@@ -727,6 +727,26 @@ def _is_new_style_class(cls):
+    else:
+        # Everything else gets a "__torch__" prefix to avoid name collisions
+        # with the names of user values.
+        module_name = "__torch__." + module_name
```
This code is missing sanity checking. I bet you can have somehow get object __name__ properties that contain '.'

zdevito commented on torch/csrc/jit/import_source.cpp:
```diff
@@ -91,88 +93,104 @@ struct SourceResolver : public Resolver {
-    std::string version_text = L.expect(TK_NUMBER).text();
+static std::unordered_set<std::string> parseImports(Lexer& L) {
+  std::unordered_set<std::string> imports;
+  while (L.cur().text() == "import") {
```
import should be a token since it is a keyword.

zdevito commented on torch/csrc/jit/import_source.cpp:
```diff
@@ -91,88 +93,104 @@ struct SourceResolver : public Resolver {
-    L.expect('=');
-    std::string version_text = L.expect(TK_NUMBER).text();
+static std::unordered_set<std::string> parseImports(Lexer& L) {
+  std::unordered_set<std::string> imports;
```
unordered_set will cause non-deterministic behavior.  This should use a vector and another thing should  avoid loading duplicates.

zdevito reviewed (commented):
This looks pretty good. I have a few suggested changes though:

* We should use the QualifiedName object for class names, moving it into c10 so that ClassType can have one. This can be done in another PR. It will avoid errors passing untyped string objects around to mean qualified names.

* We should actually use the import statements to direct how the library get loaded. It removes duplicate information (LibDefs would be hard to edit if you are hacking on the source code). But it also tests that the imports are correct. Otherwise they are just being ignored here. Implementing the import logic should not be complicated, import_print just needs a loader callback that takes a  QualifiedName and then loads the module the first time it is called with that name. That callback can be passed from the thing currently iterating over LibDefs.

The second change will change the protobuf format again, so I am wary of landing a PR and doing a followup.

suo committed:
Update on "[jit] use fully qualified name for ScriptClasses"

[jit] use fully qualified name for ScriptClasses

Make the necessary changes so that we use fully qualified names for
classes:
- Classes are registered by qualified name.
- PythonPrintPass prints qualified names for classes, and prints import
statements as well so that we can resolve those class names.
- Give ScriptTypeParser the ability to resolve qualified names to types.
- Give compiler.cpp the ability to resolve qualified names to types with
a class namespace SugaredValue.
- Change the lib export file structure to match Python, e.g. the class
foo.bar.Baz ends up in `class Baz` in foo/bar.py.

gh-metadata: pytorch pytorch 19239 gh/suo/27/head

suo committed:
Update on "[jit] use fully qualified name for ScriptClasses"

[jit] use fully qualified name for ScriptClasses

Make the necessary changes so that we use fully qualified names for
classes:
- Classes are registered by qualified name.
- PythonPrintPass prints qualified names for classes, and prints import
statements as well so that we can resolve those class names.
- Give ScriptTypeParser the ability to resolve qualified names to types.
- Give compiler.cpp the ability to resolve qualified names to types with
a class namespace SugaredValue.
- Change the lib export file structure to match Python, e.g. the class
foo.bar.Baz ends up in `class Baz` in foo/bar.py.

gh-metadata: pytorch pytorch 19239 gh/suo/27/head

suo commented:
Okay, ready for re-review. Comments that I am leaving off for the future:
1. Using QualifiedNamePtr everywhere (going to stack something right on top of this).
2. moving `op_version_set` out of source files
3. Sanity checking in the `_qualname` function

suo committed:
rebase on "[jit] use fully qualified name for ScriptClasses"

[jit] use fully qualified name for ScriptClasses

Make the necessary changes so that we use fully qualified names for
classes:
- Classes are registered by qualified name.
- PythonPrintPass prints qualified names for classes, and prints import
statements as well so that we can resolve those class names.
- Give ScriptTypeParser the ability to resolve qualified names to types.
- Give compiler.cpp the ability to resolve qualified names to types with
a class namespace SugaredValue.
- Change the lib export file structure to match Python, e.g. the class
foo.bar.Baz ends up in `class Baz` in foo/bar.py.

gh-metadata: pytorch pytorch 19239 gh/suo/27/head

suo committed:
Update on "[jit] use fully qualified name for ScriptClasses"

[jit] use fully qualified name for ScriptClasses

Make the necessary changes so that we use fully qualified names for
classes:
- Classes are registered by qualified name.
- PythonPrintPass prints qualified names for classes, and prints import
statements as well so that we can resolve those class names.
- Give ScriptTypeParser the ability to resolve qualified names to types.
- Give compiler.cpp the ability to resolve qualified names to types with
a class namespace SugaredValue.
- Change the lib export file structure to match Python, e.g. the class
foo.bar.Baz ends up in `class Baz` in foo/bar.py.

gh-metadata: pytorch pytorch 19239 gh/suo/27/head

suo committed:
Update on "[jit] use fully qualified name for ScriptClasses"

[jit] use fully qualified name for ScriptClasses

Make the necessary changes so that we use fully qualified names for
classes:
- Classes are registered by qualified name.
- PythonPrintPass prints qualified names for classes, and prints import
statements as well so that we can resolve those class names.
- Give ScriptTypeParser the ability to resolve qualified names to types.
- Give compiler.cpp the ability to resolve qualified names to types with
a class namespace SugaredValue.
- Change the lib export file structure to match Python, e.g. the class
foo.bar.Baz ends up in `class Baz` in foo/bar.py.

gh-metadata: pytorch pytorch 19239 gh/suo/27/head

suo committed:
Update on "[jit] use fully qualified name for ScriptClasses"

[jit] use fully qualified name for ScriptClasses

Make the necessary changes so that we use fully qualified names for
classes:
- Classes are registered by qualified name.
- PythonPrintPass prints qualified names for classes, and prints import
statements as well so that we can resolve those class names.
- Give ScriptTypeParser the ability to resolve qualified names to types.
- Give compiler.cpp the ability to resolve qualified names to types with
a class namespace SugaredValue.
- Change the lib export file structure to match Python, e.g. the class
foo.bar.Baz ends up in `class Baz` in foo/bar.py.

gh-metadata: pytorch pytorch 19239 gh/suo/27/head

zdevito commented on torch/csrc/jit/import_source.cpp:
```diff
@@ -105,74 +133,98 @@ struct SourceResolver : public Resolver {
-    auto name = L.expect(TK_IDENT).text();
-    L.expect('=');
-    std::string version_text = L.expect(TK_NUMBER).text();
+static std::set<std::string> parseImports(Lexer& L) {
```
Why are we importing them in a different order than they are listed in the file!? Errors will be weird!

zdevito replied:
I feel like the de-dup and sorting should happen before export, not on import.

zdevito commented on torch/csrc/jit/import_source.cpp:
```diff
@@ -105,74 +133,98 @@ struct SourceResolver : public Resolver {
-  auto& p = importer.parser_;
+    const std::vector<at::Tensor>& constant_table,
+    const std::function<void(const std::string&)>& import_callback) {
+  Parser p(src);
```
Some copypasta going on here related to imports. Can we refactor?

zdevito commented on torch/csrc/jit/passes/python_print.cpp:
```diff
@@ -1076,6 +1083,19 @@ struct PythonPrintPass {
     }
   }
 
+  std::string getImports() {
```
If we already de-dup then why do we sort and set on reimport?

zdevito reviewed (approved):
Looks good. I have a few nits around how the import process runs.

suo committed:
Update on "[jit] use fully qualified name for ScriptClasses"

[jit] use fully qualified name for ScriptClasses

Make the necessary changes so that we use fully qualified names for
classes:
- Classes are registered by qualified name.
- PythonPrintPass prints qualified names for classes, and prints import
statements as well so that we can resolve those class names.
- Give ScriptTypeParser the ability to resolve qualified names to types.
- Give compiler.cpp the ability to resolve qualified names to types with
a class namespace SugaredValue.
- Change the lib export file structure to match Python, e.g. the class
foo.bar.Baz ends up in `class Baz` in foo/bar.py.

gh-metadata: pytorch pytorch 19239 gh/suo/27/head

suo committed:
Update on "[jit] use fully qualified name for ScriptClasses"

[jit] use fully qualified name for ScriptClasses

Make the necessary changes so that we use fully qualified names for
classes:
- Classes are registered by qualified name.
- PythonPrintPass prints qualified names for classes, and prints import
statements as well so that we can resolve those class names.
- Give ScriptTypeParser the ability to resolve qualified names to types.
- Give compiler.cpp the ability to resolve qualified names to types with
a class namespace SugaredValue.
- Change the lib export file structure to match Python, e.g. the class
foo.bar.Baz ends up in `class Baz` in foo/bar.py.

gh-metadata: pytorch pytorch 19239 gh/suo/27/head
